rs/messaging/tests/memory_tests.rs

```diff
@@ -0,0 +1,500 @@
+struct Fixture {
+    pub local_env: Arc<StateMachine>,
+    pub local_canisters: Vec<CanisterId>,
+    pub remote_env: Arc<StateMachine>,
+    pub remote_canisters: Vec<CanisterId>,
+}
+
+impl Fixture {
+    /// Generates a local environment with `local_canisters_count` canisters installed;
+    /// and a remote environment with `remote_canisters_count` canisters installed.
+    fn new(config: FixtureConfig) -> Self {
+        let mut routing_table = RoutingTable::new();
+        routing_table_insert_subnet(&mut routing_table, LOCAL_SUBNET_ID).unwrap();
+        routing_table_insert_subnet(&mut routing_table, REMOTE_SUBNET_ID).unwrap();
+        let wasm = Project::cargo_bin_maybe_from_env("random-traffic-test-canister", &[]).bytes();
+
+        // Generate local environment and install canisters.
+        let local_env = StateMachineBuilder::new()
+            .with_subnet_id(LOCAL_SUBNET_ID)
+            .with_subnet_type(SubnetType::Application)
+            .with_routing_table(routing_table.clone())
+            .with_config(Some(config.local_state_machine_config()))
+            .build();
+        let local_canisters = (0..config.local_canisters_count)
+            .map(|_| install_canister(&local_env, wasm.clone()))
+            .collect();
+
+        // Generate remote environment and install canisters.
+        let remote_env = StateMachineBuilder::new()
+            .with_subnet_id(REMOTE_SUBNET_ID)
+            .with_subnet_type(SubnetType::Application)
+            .with_routing_table(routing_table.clone())
+            .with_config(Some(config.remote_state_machine_config()))
+            .build();
+        let remote_canisters = (0..config.remote_canisters_count)
+            .map(|_| install_canister(&remote_env, wasm.clone()))
+            .collect();
+
+        Self {
+            local_env: local_env.into(),
+            local_canisters,
+            remote_env: remote_env.into(),
+            remote_canisters,
+        }
+    }
+
+    /// Returns all canisters installed in the fixture, local canisters first.
+    pub fn canisters(&self) -> Vec<CanisterId> {
+        self.local_canisters
+            .iter()
+            .cloned()
+            .chain(self.remote_canisters.iter().cloned())
+            .collect()
+    }
+
+    /// Returns a reference to `self.local_env` if `canister` is installed on it.
+    /// Else a reference to `self.remote_env` if `canister` is installed on it.
+    ///
+    /// Panics if `canister` is not installed on either env.
+    pub fn get_env(&self, canister: &CanisterId) -> &StateMachine {
+        if self.local_canisters.contains(canister) {
+            return &self.local_env;
+        }
+        if self.remote_canisters.contains(canister) {
+            return &self.remote_env;
+        }
+        unreachable!();
+    }
+
+    /// Helper function for replacing canister state elements.
+    ///
+    /// Panics if `canister` is not installed in `Self`.
+    fn replace_canister_state<T>(
```

Comment:
Personal preference: I would have named all of the `replace_` methods (here and in the canister) `set_`. It's more standard.

Also, this particular method does not actually replace anything. It calls a given canister's method with the given payload (and assumes that the return type is the same as the payload).

Reply:
I had `set_` at first but with calls that really only set and not return the previous value. During debugging I changed that to also return the previous value for which I used [replace](https://doc.rust-lang.org/std/thread/struct.LocalKey.html#method.replace), so I imported the name here. By contrast there is also [set](https://doc.rust-lang.org/std/thread/struct.LocalKey.html#method.set), which doesn't return the previous value.

We can use our own nomenclature; I'm fine with either one - changed.